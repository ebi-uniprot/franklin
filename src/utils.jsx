import React from 'react';

export function getLastIndexOfSubstringIgnoreCase(string, substring) {
  return string.toLowerCase().lastIndexOf(substring.toLowerCase());
}

export const getFlattenedPaths = (currentItems, id, path = []) => {
  let flattened = [];
  currentItems.forEach((node) => {
    const { items, ...thisNode } = node;
    const nodePath = [...path, thisNode];
    if (items) {
      const result = getFlattenedPaths(items, id, nodePath);
      if (result.length) {
        flattened = [...flattened, ...result];
      }
    } else if (!id || thisNode.id === id) {
      flattened = [...flattened, nodePath];
    }
  });
  return flattened;
};

export function restructureFlattenedTreeItemsForAutocomplete(
  items,
  sep = ' / '
) {
  return {
    id: items[items.length - 1].id,
    pathLabel: items.map((item) => item.label).join(sep),
    itemLabel: items[items.length - 1].label,
  };
}

export function restructureFlattenedTreeDataForAutocomplete(flattenedTreeData) {
  return flattenedTreeData.map((items) =>
    restructureFlattenedTreeItemsForAutocomplete(items)
  );
}

export function formatLargeNumber(x) {
  return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

export function capitaliseFirstLetter(string) {
  return string && string.charAt(0).toUpperCase() + string.slice(1);
}

export function highlightSubstring(string, substring) {
  const i = getLastIndexOfSubstringIgnoreCase(string, substring);
  if (i < 0) return string;
  const prestring = string.slice(0, i);
  const highlight = string.slice(i, i + substring.length);
  const poststring = string.slice(i + substring.length);
  return (
    <>
      {prestring}
      <b>{highlight}</b>
      {poststring}
    </>
  );
}

export function getClassNames(mixedInput, optionalCondition, output = '') {
  if (typeof optionalCondition !== 'undefined' && optionalCondition !== true) {
    return '';
  }

  if (typeof mixedInput === 'string') {
    return mixedInput;
  }

  if (Array.isArray(mixedInput)) {
    if (typeof mixedInput[1] === 'boolean') {
      return getClassNames(mixedInput[0], mixedInput[1], output);
    }

    return mixedInput
      .map((item) => getClassNames(item, optionalCondition, output))
      .filter((x) => x) // removes nulls, undefineds, empty strings, etc.
      .join(' ');
  }

  return output;
}
