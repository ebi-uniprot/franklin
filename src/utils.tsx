export function getLastIndexOfSubstringIgnoreCase(
  string: string,
  substring: string
) {
  return string.toLowerCase().lastIndexOf(substring.toLowerCase());
}

export type Item = {
  label: string;
  id: string;
  items?: Item[];
};

export const getFlattenedPaths = (
  currentItems: Item[],
  id?: string,
  path: Item[] = []
) => {
  let flattened: Pick<Item, 'label' | 'id'>[][] = [];
  currentItems.forEach((node) => {
    const { items, ...thisNode } = node;
    const nodePath = [...path, thisNode];
    if (items) {
      const result = getFlattenedPaths(items, id, nodePath);
      if (result.length) {
        flattened = [...flattened, ...result];
      }
    } else if (!id || thisNode.id === id) {
      flattened = [...flattened, nodePath];
    }
  });
  return flattened;
};

export function restructureFlattenedTreeItemsForAutocomplete(
  items: Item[],
  sep = ' / '
) {
  return {
    id: items[items.length - 1].id,
    pathLabel: items.map((item) => item.label).join(sep),
    itemLabel: items[items.length - 1].label,
  };
}

export function restructureFlattenedTreeDataForAutocomplete(
  flattenedTreeData: Item[][]
) {
  return flattenedTreeData.map((items) =>
    restructureFlattenedTreeItemsForAutocomplete(items)
  );
}

export function formatLargeNumber(x: string | number) {
  return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

export function capitaliseFirstLetter(string?: string | null) {
  return string && string.charAt(0).toUpperCase() + string.slice(1);
}

export function highlightSubstring(string: string, substring: string) {
  const i = getLastIndexOfSubstringIgnoreCase(string, substring);
  if (i < 0) return string;
  const prestring = string.slice(0, i);
  const highlight = string.slice(i, i + substring.length);
  const poststring = string.slice(i + substring.length);
  return (
    <>
      {prestring}
      <b>{highlight}</b>
      {poststring}
    </>
  );
}

const reProtocol = /^(https?:)?(\/\/)?/;
const reTrailingSlashes = /(\/+$)/;

export const tidyUrlString = (url: string) =>
  url.replace(reProtocol, '').replace(reTrailingSlashes, '');
