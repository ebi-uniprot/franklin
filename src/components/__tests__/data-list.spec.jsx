import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import DataList from '../data-list';
import { fillArray } from '../../utils';

describe('DataList', () => {
  const data = fillArray(10, (_, index) => ({
    id: `id${index}`,
    content: index,
  }));

  let onLoadMoreItems;
  const scrollDataAttribute = 'data-list';
  const getRender = (hasMoreData = true) =>
    render(
      <div
        style={{ height: '65vh', overflowY: 'auto' }}
        data-loader-scroll={scrollDataAttribute}
        data-testid="scroll-container"
      >
        <DataList
          onLoadMoreItems={onLoadMoreItems}
          hasMoreData={hasMoreData}
          data={data}
          dataRenderer={(item) => <p>{item.content}</p>}
          onCardClick={null}
          scrollDataAttribute={scrollDataAttribute}
        />
      </div>
    );
  beforeEach(() => {
    onLoadMoreItems = jest.fn();
  });

  test('should render', () => {
    const { asFragment } = getRender();
    expect(asFragment()).toMatchSnapshot();
  });

  test('should request more data', () => {
    const { getByTestId } = getRender();
    const scrollContainer = getByTestId('scroll-container');
    fireEvent.scroll(scrollContainer, { target: { scrollY: 1000 } });
    expect(onLoadMoreItems).toHaveBeenCalled();
  });

  test('should not request more data', () => {
    const { getByTestId } = getRender(false);
    const scrollContainer = getByTestId('scroll-container');
    fireEvent.scroll(scrollContainer, { target: { scrollY: 1000 } });
    expect(onLoadMoreItems).toHaveBeenCalledTimes(0);
  });
});
