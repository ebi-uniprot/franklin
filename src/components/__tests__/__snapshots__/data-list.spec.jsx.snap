// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataList should render 1`] = `
<DocumentFragment>
  <div
    data-loader-scroll="data-list"
    data-testid="scroll-container"
    style="height: 65vh; overflow-y: auto;"
  >
    <section>
      <p>
        0
      </p>
    </section>
    <section>
      <p>
        1
      </p>
    </section>
    <section>
      <p>
        2
      </p>
    </section>
    <section>
      <p>
        3
      </p>
    </section>
    <section>
      <p>
        4
      </p>
    </section>
    <section>
      <p>
        5
      </p>
    </section>
    <section>
      <p>
        6
      </p>
    </section>
    <section>
      <p>
        7
      </p>
    </section>
    <section>
      <p>
        8
      </p>
    </section>
    <section>
      <p>
        9
      </p>
    </section>
    <div
      class="data-loader__loading"
    >
      Loading...
    </div>
  </div>
</DocumentFragment>
`;
